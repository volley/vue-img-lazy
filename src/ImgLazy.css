img.img-lazy:not([src]) {
    visibility: hidden;
}
