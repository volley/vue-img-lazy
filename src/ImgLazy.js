import Vue from 'vue';
import LazyLoad from 'vanilla-lazyload';
import './ImgLazy.css';

export default {
  name: 'ImgLazy',
  functional: true,
  props: {
    src: {
      type: String,
      required: true,
    },
    srcset: {
      type: String,
      required: false,
      default: undefined,
    },
    sizes: {
      type: String,
      required: false,
      default: undefined,
    },
    alt: {
      type: String,
      required: false,
      default: undefined,
    },
    width: {
      type: String,
      required: false,
      default: undefined,
    },
    height: {
      type: String,
      required: false,
      default: undefined,
    },
  },
  render(createElement, { props, data }) {
    // single global instance
    if (!document.imgLazyLoadInstance) {
      // See https://github.com/verlok/lazyload#options for what these options are.
      document.imgLazyLoadInstance = new LazyLoad({
        element_selector: '.img-lazy', // css selector for our class
        threshold: 300, // load images when within 300px of the viewport
      });
    }

    // schedule update on next tick after our new img element has been rendered (on this tick)
    Vue.nextTick().then(() => {
      document.imgLazyLoadInstance.update();
    });

    return createElement('img', {
      ...data,

      // always override with a key based on props so this DOM element is completely replaced
      // and re-lazy'd when altered in any way. Otherwise the lazyload library ignores the element
      // even when changes occur because it has already marked it as processed.
      key: Object.values(props).join(','),

      attrs: {
        ...data.attrs,
        alt: props.alt,
        'data-src': props.src,
        'data-srcset': props.srcset,
        'data-sizes': props.sizes,
        width: props.width,
        height: props.height,
      },
      class: {
        ...data.class,
        'img-lazy': true,
      },
    });
  },
};
