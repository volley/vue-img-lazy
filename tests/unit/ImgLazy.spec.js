import "bdd-lazy-var/global";
import { mount } from "@vue/test-utils";
import ImgLazy from "../../src/ImgLazy";

describe("ImgLazy", () => {
  def('extraAttrs', () => ({}));

  subject(() =>
    mount(ImgLazy, {
      context: {
        attrs: $extraAttrs,
        props: {
          src: "http://test.host/myimg.jpg",
          srcset:
            "http://test.host/myimg.jpg http://test.host/myimg_big.jpg x2",
          sizes: "1 2 3",
          alt: "super cool pictures",
          width: "100",
          height: "100",
        }
      }
    })
  );

  it("renders this, and you changed it on purpose if it did not", () => {
    expect($subject).toMatchSnapshot(`html`);
  });

  it("should set a Vue key composed of the props so that Vue knows not to recycle this node if props change", () => {
    expect($subject.rootNode.key).toMatchInlineSnapshot(
      `"http://test.host/myimg.jpg,http://test.host/myimg.jpg http://test.host/myimg_big.jpg x2,1 2 3,super cool pictures,100,100"`
    );
  });

  describe('with crossorigin attribute', function () {
    def('extraAttrs', () => ({
      crossorigin: 'anonymous',
    }));

    it("adds crossorigin attribute", () => {
      expect($subject.element.getAttribute('crossorigin')).toEqual('anonymous');
    });
  });
});
