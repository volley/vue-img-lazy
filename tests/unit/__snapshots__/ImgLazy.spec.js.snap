// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImgLazy renders this, and you changed it on purpose if it did not: html 1`] = `<img alt="super cool pictures" data-src="http://test.host/myimg.jpg" data-srcset="http://test.host/myimg.jpg http://test.host/myimg_big.jpg x2" data-sizes="1 2 3" width="100" height="100" class="img-lazy">`;
